File: ai-solutions/04/gemini-3-pro/python/main.py
import sys
import os

def solve():
    # Default path relative to this script
    input_path = os.path.join(os.path.dirname(__file__), "../../../../inputs/04.txt")
    if len(sys.argv) > 1:
        input_path = sys.argv[1]
    
    try:
        with open(input_path, 'r') as f:
            grid = [list(line.strip()) for line in f if line.strip()]
    except FileNotFoundError:
        print(f"Error: Input file not found at {input_path}")
        return

    rows = len(grid)
    cols = len(grid[0])

    def count_neighbors(r, c, current_grid):
        count = 0
        for dr in [-1, 0, 1]:
            for dc in [-1, 0, 1]:
                if dr == 0 and dc == 0:
                    continue
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols:
                    if current_grid[nr][nc] == '@':
                        count += 1
        return count

    # Part 1
    p1_count = 0
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '@':
                if count_neighbors(r, c, grid) < 4:
                    p1_count += 1
    
    print(p1_count)

    # Part 2
    # Create a working copy of the grid
    current_grid = [row[:] for row in grid]
    total_removed = 0
    
    while True:
        to_remove = []
        for r in range(rows):
            for c in range(cols):
                if current_grid[r][c] == '@':
                    if count_neighbors(r, c, current_grid) < 4:
                        to_remove.append((r, c))
        
        if not to_remove:
            break
            
        total_removed += len(to_remove)
        for r, c in to_remove:
            current_grid[r][c] = '.'
            
    print(total_removed)

if __name__ == "__main__":
    solve()

File: ai-solutions/04/gemini-3-pro/rust/src/main.rs
use std::env;
use std::fs;
use std::path::PathBuf;

fn main() {
    let args: Vec<String> = env::args().collect();
    
    // Default path logic is a bit tricky with compiled binaries.
    // We'll try to resolve relative to the current working directory if no arg is provided.
    // Assuming running from ai-solutions/04/gemini-3-pro/rust via 'cargo run'
    let default_path = PathBuf::from("../../../../inputs/04.txt");
    
    let input_path = if args.len() > 1 {
        PathBuf::from(&args[1])
    } else {
        default_path
    };

    let contents = fs::read_to_string(&input_path).expect(&format!("Something went wrong reading the file at {:?}", input_path));
    let grid: Vec<Vec<char>> = contents
        .lines()
        .filter(|line| !line.trim().is_empty())
        .map(|line| line.trim().chars().collect())
        .collect();

    let rows = grid.len();
    if rows == 0 {
        return;
    }
    let cols = grid[0].len();

    // Part 1
    let mut p1_count = 0;
    for r in 0..rows {
        for c in 0..cols {
            if grid[r][c] == '@' {
                if count_neighbors(&grid, r, c) < 4 {
                    p1_count += 1;
                }
            }
        }
    }
    println!("{}", p1_count);

    // Part 2
    let mut current_grid = grid.clone();
    let mut total_removed = 0;

    loop {
        let mut to_remove = Vec::new();
        for r in 0..rows {
            for c in 0..cols {
                if current_grid[r][c] == '@' {
                    if count_neighbors(&current_grid, r, c) < 4 {
                        to_remove.push((r, c));
                    }
                }
            }
        }

        if to_remove.is_empty() {
            break;
        }

        total_removed += to_remove.len();
        for (r, c) in to_remove {
            current_grid[r][c] = '.';
        }
    }
    println!("{}", total_removed);
}

fn count_neighbors(grid: &Vec<Vec<char>>, r: usize, c: usize) -> usize {
    let rows = grid.len() as isize;
    let cols = grid[0].len() as isize;
    let mut count = 0;

    for dr in -1..=1 {
        for dc in -1..=1 {
            if dr == 0 && dc == 0 {
                continue;
            }
            let nr = r as isize + dr;
            let nc = c as isize + dc;

            if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
                if grid[nr as usize][nc as usize] == '@' {
                    count += 1;
                }
            }
        }
    }
    count
}

